### PR TITLE
[SPARK-22224][SQL]Override toString of KeyValue/Relational-GroupedDataset

### PR DESCRIPTION
**Pull request by yaooqinn**
#### before

```scala
scala> val words = spark.read.textFile("README.md").flatMap(_.split(" "))
words: org.apache.spark.sql.Dataset[String] = [value: string]

scala> val grouped = words.groupByKey(identity)
grouped: org.apache.spark.sql.KeyValueGroupedDataset[String,String] = org.apache.spark.sql.KeyValueGroupedDataset@65214862
```
#### after
```scala
scala> val words = spark.read.textFile("README.md").flatMap(_.split(" "))
words: org.apache.spark.sql.Dataset[String] = [value: string]

scala> val grouped = words.groupByKey(identity)
grouped: org.apache.spark.sql.KeyValueGroupedDataset[String,String] = [key: [value: string], value: [value: string]]
```

## How was this patch tested?
existing ut

cc @gatorsmile @cloud-fan

### CODE REVIEW

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 1]
```diff
@@ -54,6 +55,14 @@ class KeyValueGroupedDataset[K, V] private[sql](
   private def sparkSession = queryExecution.sparkSession
 
   /**
+   * Returns the schema of this Dataset.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  def schema: StructType = queryExecution.analyzed.schema
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Can you remove this?

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: ok

**Comment by yaooqinn** [comment]
cc again @cloud-fan @gatorsmile

**Comment by gatorsmile** [comment]
cc @viirya Could you review this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 2]
```diff
@@ -564,4 +565,30 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
+        builder.append(" ... " + (kFields.length - 2) + " more field(s)")
+      }
+      builder.append("], value: [")
+      builder.append(vFields.take(2).mkString(", "))
+      if (vFields.length > 2) {
+        builder.append(" ... " + (vFields.length - 2) + " more field(s)")
+      }
+      builder.append("]]").toString()
+    } catch {
+      case NonFatal(e) =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: When we will encounter this error?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 3]
```diff
@@ -18,16 +18,17 @@
...
 import org.apache.spark.annotation.{Experimental, InterfaceStability}
 import org.apache.spark.api.java.function._
 import org.apache.spark.sql.catalyst.encoders.{encoderFor, ExpressionEncoder}
 import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, CreateStruct}
 import org.apache.spark.sql.catalyst.plans.logical._
-import org.apache.spark.sql.catalyst.streaming.InternalOutputModes
 import org.apache.spark.sql.execution.QueryExecution
 import org.apache.spark.sql.expressions.ReduceAggregator
 import org.apache.spark.sql.streaming.{GroupState, GroupStateTimeout, OutputMode}
+import org.apache.spark.sql.types.StructType
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Why import `StructType`? I didn't see you use it.

**Comment by viirya** [comment]
Maybe add a simple test?

**Comment by viirya** [comment]
Although this `toString` in scala-shell looks good, when you print out directly, it might look weird because you just see:

```scala
[key: [value: string], value: [value: string]]
```

It's better to add prefix of the class name, like.

```scala
KeyValueGroupedDataset: [key: [value: string], value: [value: string]]
```

**Comment by viirya** [comment]
Btw, I think this might need a minor/trivial JIRA ticket.

**Comment by cloud-fan** [comment]
shall we fix RelationalGroupedDataset too?

**Comment by yaooqinn** [comment]
Ok，i will add a jira tgt and fix RelationalGroupedDataset

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: same question

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: I once thought you would answer that for me https://github.com/apache/spark/blame/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L374, LOL

**Comment by viirya** [comment]
Please replace `[Minor]` with `[SQL]`.

**Comment by yaooqinn** [comment]
cc @viirya @cloud-fan @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Although some methods of `StringBuilder` can throw exceptions, I didn't see simple `append` can throw exception. Access of `schema` should not throw exception at all. So I don't get why it is.

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 4]
```diff
@@ -465,6 +466,24 @@ class RelationalGroupedDataset protected[sql](
 
     Dataset.ofRows(df.sparkSession, plan)
   }
+
+  override def toString: String = {
+    try {
+      val builder = new StringBuilder
+      builder.append("RelationalGroupedDataset: [key: [")
+      val kFields = groupingExprs.map(_.asInstanceOf[NamedExpression]).map {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: Are all expressions in `groupingExprs` `NamedExpression`?

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 5]
```diff
@@ -465,6 +466,24 @@ class RelationalGroupedDataset protected[sql](
 
     Dataset.ofRows(df.sparkSession, plan)
   }
+
+  override def toString: String = {
+    try {
+      val builder = new StringBuilder
+      builder.append("RelationalGroupedDataset: [key: [")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: `key:` or `groupingBy:`?

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 6]
```diff
@@ -465,6 +466,24 @@ class RelationalGroupedDataset protected[sql](
...
+      val kFields = groupingExprs.map(_.asInstanceOf[NamedExpression]).map {
+        case f => s"${f.name}: ${f.dataType.simpleString(2)}"
+      }
+      builder.append(kFields.take(2).mkString(", "))
+      if (kFields.length > 2) {
+        builder.append(" ... " + (kFields.length - 2) + " more field(s)")
+      }
+      builder.append(s"], value: ${df.toString}, $groupType]").toString()
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: `value:` or `df:`?

`$groupType` -> `type: $groupType`?

File: sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala [hunk 7]
```diff
@@ -46,6 +36,20 @@ abstract class QueryTest extends PlanTest {
   Locale.setDefault(Locale.US)
 
   /**
+   * Makes sure the answer matches the expected result
+   */
+  def checkString(expected: String, actual: String): Unit = {
+    if (expected != actual) {
+      fail(
+        "KeyValueGroupedDataset.toString() gives wrong result:\n\n" + sideBySide(
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala]: You didn't only use this against `KeyValueGroupedDataset`.

File: sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.test.SharedSQLContext
+
+class RelationalGroupedDatasetSuite extends QueryTest with SharedSQLContext {
+  import testImplicits._
+
+  test("Check RelationalGroupedDataset toString: Single data") {
+    val kvDataset = (1 to 3).toDF("id").as[SingleData].groupBy("id")
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala]: Is `SingleData` necessary?

File: sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    val expected = "RelationalGroupedDataset: [key: [id: int], value: [id: int], GroupByType]"
+    val actual = kvDataset.toString
+    checkString(expected, actual)
+  }
+
+  test("Check RelationalGroupedDataset toString: over length schema ") {
+    val kvDataset = (1 to 3).map( x => (x, x.toString, x.toLong))
+      .toDF("id", "val1", "val2").as[TripleData].groupBy("id")
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala]: Is `TripleData` necessary?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 10]
```diff
@@ -18,13 +18,13 @@
 package org.apache.spark.sql
 
 import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: unnecessary import

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 11]
```diff
@@ -21,6 +21,7 @@ import java.util.Locale
 
 import scala.collection.JavaConverters._
 import scala.language.implicitConversions
+import scala.util.control.NonFatal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 12]
```diff
@@ -465,6 +466,19 @@ class RelationalGroupedDataset protected[sql](
 
     Dataset.ofRows(df.sparkSession, plan)
   }
+
+  override def toString: String = {
+    val builder = new StringBuilder
+    builder.append("RelationalGroupedDataset: [groupingBy: [")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: `groupingBy` -> `group by`

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 13]
```diff
@@ -465,6 +466,19 @@ class RelationalGroupedDataset protected[sql](
...
+    val kFields = groupingExprs.map(_.asInstanceOf[NamedExpression]).map {
+      case f => s"${f.name}: ${f.dataType.simpleString(2)}"
+    }
+    builder.append(kFields.take(2).mkString(", "))
+    if (kFields.length > 2) {
+      builder.append(" ... " + (kFields.length - 2) + " more field(s)")
+    }
+    builder.append(s"], df: ${df.toString}, type: $groupType]").toString()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: why include `df` here?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 14]
```diff
@@ -564,4 +564,25 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
+    builder.append(kFields.take(2).mkString(", "))
+    if (kFields.length > 2) {
+      builder.append(" ... " + (kFields.length - 2) + " more field(s)")
+    }
+    builder.append("], value: [")
+    builder.append(vFields.take(2).mkString(", "))
+    if (vFields.length > 2) {
+      builder.append(" ... " + (vFields.length - 2) + " more field(s)")
+    }
+    builder.append("]]").toString()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: shall we include `df` here?

File: sql/core/src/test/scala/org/apache/spark/sql/KeyValueGroupedDatasetSuite.scala [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.test.SharedSQLContext
+
+class KeyValueGroupedDatasetSuite extends QueryTest with SharedSQLContext {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/KeyValueGroupedDatasetSuite.scala]: we can put the tests in `DataSetSuite`, as `DataSet.toString` is also tested there.

File: sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.test.SharedSQLContext
+
+class RelationalGroupedDatasetSuite extends QueryTest with SharedSQLContext {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/RelationalGroupedDatasetSuite.scala]: we can put the tests in `DataSetSuite`, as `DataSet.toString` is also tested there.

File: sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala [hunk 17]
```diff
@@ -46,6 +36,20 @@ abstract class QueryTest extends PlanTest {
   Locale.setDefault(Locale.US)
 
   /**
+   * Makes sure the answer matches the expected result
+   */
+  def checkString(expected: String, actual: String): Unit = {
+    if (expected != actual) {
+      fail(
+        "The actual string gives wrong result:\n\n" + sideBySide(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/QueryTest.scala]: it's kind of an overkill, the `toString` we are testing always return a single line, just `assert` is enough.

**Comment by yaooqinn** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 18]
```diff
@@ -1341,8 +1341,69 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       Seq(1).toDS().map(_ => ("", TestForTypeAlias.seqOfTupleTypeAlias)),
       ("", Seq((1, 1), (2, 2))))
   }
+
+  test("Check RelationalGroupedDataset toString: Single data") {
+    val kvDataset = (1 to 3).toDF("id").groupBy("id")
+    val expected = "RelationalGroupedDataset: [" +
+      "group by: [id: int], value: [id: int], type: GroupByType]"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: a better idea:
* `group by` -> `grouping expressions`
* `type: GroupByType` -> `type: GroupBy`, i.e. remove the `Type` suffix.

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 19]
```diff
@@ -479,7 +492,9 @@ private[sql] object RelationalGroupedDataset {
   /**
    * The Grouping Type
    */
-  private[sql] trait GroupType
+  private[sql] trait GroupType {
+    override def toString: String = getClass.getSimpleName.replace("$", "")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: `getClass.getSimpleName.stripSuffix("$").stripSuffix("Type")`

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by yaooqinn** [comment]
hi @cloud-fan jenkins ok

**Comment by cloud-fan** [comment]
merging to master!

**Pull request closed without merging**